distage/distage-testkit-core/src/main/scala/izumi/distage/testkit/services/dstest/DistageTestRunner.scala

```diff
@@ -73,10 +74,17 @@ class DistageTestRunner[F[_]: TagK]
         // here we find all the shared components in each of our individual tests
         val sharedKeys = testPlans.map(_._2).flatMap {
           plan =>
-            plan.steps.filter(env memoizationRoots _.target).map(_.target)
+            val memoized = plan.steps.filter(env memoizationRoots _.target).map(_.target)
+            val weak = plan.steps.collect {
+              // here we adding each weak reference into shared keys (we assume that if we have a weak reference in one test plan - it's should be bind)
+              case w: WiringOp.ReferenceKey if w.wiring.weak => w.target
+              // adding sets into shared keys if one of the members is preserving in test plan (we should build set with a weak reference before tests start)
+              case c: CreateSet if c.members.exists(plan.index.keys.toList.contains) => c.target
```

Comment:
I think you can't solve the problem by adding roots, because roots are _new components_ that result in new large subgraphs being added to shared locator. If you add all sets with weak members into roots, the result graph will be much-much larger than the original graph of "only dependencies of memoizationRoots"

I think what needs to be done is pass a custom weakness predicate to `TracingDIGC`,
such that **during GC** when it encounters a weak key it may be considered referenced, when it's present in test plans. This would make it include ONLY those weak references that are in sets that ARE dependencies of memoizationRoots, not ALL sets.

Reply:
Hm, okay then, I'll take a look into.
I've thought that it would be okay to store in memoizationRoots whole set of weak references.

Reply:
It's 'okay', but it won't behave as expected. e.g. when you have memoizationRoots=empty your weak DDLs will still be memoized, and with them everything that depends on them - i.e. a huge chunk of the graph. You'd expect that `memoizationRoots=Set.empty` means no memoization at all, but this is not the case when additional roots are added synthetically...

Reply:
@neko-kai may you check new implementation ?